Multi-resource hook for fetching permissions

davidlougheed opened the pull request:
needs #25 first; breaking change

davidlougheed committed:
feat: multi-resource hook for fetching permissions
BREAKING CHANGE: fetch resource permissions action requires array

davidlougheed committed:
Merge remote-tracking branch 'origin/main' into feat/auth/multi-resource-perms

# Conflicts:
#	src/redux/authSlice.ts

davidlougheed committed:
refact: factor out selector into a useAuthState hook

davidlougheed committed:
chore: rename FetchPermissionsPayload type
BREAKING CHANGE: type name change (pluralization)

davidlougheed committed:
chore: fix type hint for recursiveOrderedObject

davidlougheed committed:
Merge remote-tracking branch 'origin/main' into feat/auth/multi-resource-perms

davidlougheed commented on src/redux/authSlice.ts:
```diff
@@ -102,31 +102,36 @@ export const refreshTokens = createAsyncThunk<RefreshTokenPayload, string>(
+            return resources.reduce((acc, resource) => {
+                const key = makeResourceKey(resource);
+                const rp = auth.resourcePermissions?.[key];
+                return acc || !rp?.isFetching;
```
this should probably be an &&... tbd.

v-rocheleau replied:
Yeah probably, would also be better to use `every` instead of `reduce` for faster condition eval

v-rocheleau commented on src/hooks.ts:
```diff
@@ -16,47 +16,67 @@ const AUTH_RESULT_TYPE = "authResult";
     useEffect(() => {
-        if (!haveAuthorizationService || isFetching || permissions || hasAttempted) return;
-        dispatch(fetchResourcePermissions({ resource, authzUrl }));
+        const allFetching = keys.reduce((acc, key) => acc || !!resourcePermissions[key]?.isFetching, false);
+        const allHavePermissions =
+            keys.reduce((acc, key) => acc && !!resourcePermissions[key]?.permissions?.length, true);
+        const allAttempted = keys.reduce((acc, key) => acc && !!resourcePermissions[key]?.hasAttempted, true);
```
I think using `Array.every()` instead of `Array.reduce()` would be better here, since `every` will exit as soon as the condition is not satisfied.

davidlougheed replied:
forgot about that one - gotta change some WIP bento web stuff too...

davidlougheed replied:
Array.some is also good elsewhere in this diff

v-rocheleau commented on src/redux/authSlice.ts:
```diff
@@ -258,37 +263,45 @@ export const authSlice = createSlice({
-                    permissions: payload?.result?.[0] ?? [],
-                };
+            .addCase(fetchResourcesPermissions.fulfilled, (state, { meta, payload }) => {
+                const resources = meta.arg.resources;
+                for (const r in resources) {
+                    const key = makeResourceKey(resources[r]);
```
Make uniform with the other loops

davidlougheed replied:
like with the resources var? in this case its done differently because we index into the payload + access `resources` more than once

v-rocheleau replied:
Oh didn't notice sorry, all good then

davidlougheed committed:
perf: use .some/.every for boolean consolidation
also swaps conditions for fetching permissions to make sure nothing is fetching
that would lead to a race condition.

v-rocheleau reviewed (approved):
lgtm
